### ISSUE TITLE
Zephyr's net stack does not handle fragmentation of packets properly

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#174]
**_Reported by Gustavo Lima Chaves:_**

One more of our evidences that the 6LOWPAN fragmentation code in Zephyr seems broken is due to the way said stack was adapted to work with multiple network buffers. We came across this issue by testing a board with Zephyr against one with RIOT, and since the latter has IPHC partially broken, it had to be disabled on both ends to allow them to communicate, which obviously resulted in the network packets being large enough to not fit within the radio MTU. After comparing to the original Contiki code and going through the RFC for both 6LOWPAN and 6LOWPAN-IPHC, and also looking at the Wireshark log, it became clear Zephyr is not respecting the way fragmentation should work. After the user sends a packet, it goes through the Contiki stack, where the UDP and IPv6 headers will be set, before sending that net_buf handle back to Zephyr so it can go through the MAC layer. It's at this point that things start to break, as the original Contiki code is now spread across several components, among them 'compression' and 'fragmentation'. According to the RFC, the fragmentation header should include the original size of the packet before compression, but by the time it reaches this point, it has already been compressed (or not when IPHC is not used, but
even that changed the original buffer by adding the 6LOWPAN IPV6_DISPATCH byte), and because of how the code has been adapted to the multiple buffers, the original information appears to be lost by now, and the fragmentation header contains wrong information. Not only that, but also the semantics of the code doing the fragmentation still work as they did on the original code, so they payload itself ends up being of the wrong size. Even if the reassembly code were to account for these inconsistencies, other implementations would not be able to communicate correctly with Zephyr.

(Imported from Jira ZEP-208)

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

* [test.tgz](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-174/test.tgz) (size: 2602)

**Comment by nashif** [comment]
**_by Tomasz Bursztyka:_**

Title is wrong: cc2520 is below the IP stack, and the driver is not in charge of handling fragmentation (that's the job of the 802.15.4 stack witch is included in net stack).

And the bug is not exactly a bug. 802.15.4 has no meaning to be a reliable link, there is no reliability means into it (unlike 802.11 which has mechanism for re-transmission etc... ).
However on Quark SE devboard, the integrated cc2520 makes it really bad: on our own test we get about ~25-30% of packet lost that way (usually badly received: FCS check fails, thus the bad CRC. It's more rare to see a bad transmitted packet: a sniffer + wireshark helps to see that).
The chip itself is not the culprit here, most probably the antenna for some reasons. 

Anyway, to achieve a much better "reliability", I made a patch last week which adds proper support for ACK request/answer handling. It is a mechanism being part of 802.15.4, but not meant to be enabled for all packets. Using it, we can tell most of the time if a packet was properly received or not, and thus re-transmit it if necessary (this is not in cc2520's driver, it's in the RDC driver in net stack)

Just:
CONFIG_NETWORKING_WITH_15_4_ALWAYS_ACK=y
and on cc2520 side (it's not mandatory, but use it: it will handle ACK answers automatically from the hw thus less job for the host cpu/os):
CONFIG_TI_CC2520_AUTO_ACK=y

This way, the boards will request an ACK for each data packet sent (broadcast packet can't request an ACK obviously), and the other end will send an ACK if only it finds out the FCS being fine. Earlier tests made on that board (not by me and not on Zephyr but with contiki) have used exactly the same trick, and as far as I know there is nothing else doable to improve that on this board. I bet using cc2520 daughter board on which you can plug a proper antenna would work way better.

Try this, and let us know.

**Comment by nashif** [comment]
**_by Ravi kumar Veeramally:_**

can you point to me the rfc details about "the fragmentation header should include the original size of the packet before compression".

**Comment by nashif** [comment]
**_by Tomasz Bursztyka:_**

Actually, if you guys have found a bug on the compression/fragmentation: please open a specific bug ticket for it. Let's not mix different things in the same ticket.

**Comment by nashif** [comment]
**_by Ravi kumar Veeramally:_**

https://tools.ietf.org/html/rfc6282
senction 2:
   Section 5.3 of [RFC4944] defines the fragment header's datagram_size and
   datagram_offset values as the size and offset of the IPv6 datagram
   before compression.

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

I can reword the title as you wish. I also know the driver's code may not be the culprit, I just have a test involving it, thus the citation. The *very same board* has been made to work with Contiki (plus Soletta), transmitting packets of arbitrary size to another board running RIOT with success (only IPHC turned off). I will test the ack part of the WIP of yours, but that seems to have nothing to do with the fragmentation issues we get.

**Comment by nashif** [comment]
**_by Tomasz Bursztyka:_**

At least very the first part of the bug descriptions looks really much like the issue I know of (getting a lot of Bad CRC error).

Anyway, about compression, netcontiki uses HC06 not IPHC. Maybe the issue comes from it, I don't know (never touched that side of the stack myself)

**Comment by nashif** [comment]
**_by Iván Briano:_**

Yes, my comment that Gustavo quoted here has nothing to do with the radio driver or hardware or the reliability of this type of link. In fact, when I was testing this and found the problem, the radio link was remarkably stable, with a very low rate of packets caught by the sniffer with a bad FCS.

The fragmentation part is on the stack, not the radio, and it's wrong. As Ravi mentioned, the RFC states that the datagram size should be before compression. Another thing I noted is that when the 6LoWPAN dispatch headers are added (with or without compression), they both increment the size of the packet that ends up used to set the datagram size, so even when there is no compression, fragmentation is wrong. It accounts for the dispatch header as part of the payload when it shouldn't and ends up sending one byte less than it should. You can see this with a sniffer by checking the packet's payload size being one byte less than expected (should always be multiples of 8). The next packet will contain these byte, but still have the wrong datagram size, and with its offset correct, so in the reassembly there'd end up being a whole.

**Comment by nashif** [comment]
**_by Iván Briano:_**

Tomasz, I was disabling all forms of compression explicitly because I was testing a Zephyr server against a RIOT client, and IPHC on RIOT is partially broken so it kept dropping packets. The fragmentation problem exists regardless of compression mode used.

**Comment by nashif** [comment]
**_by Tomasz Bursztyka:_**

Ok thanks for the precisions, it's clearer.

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

Thanks for input as well, Tomasz. I'm now testing more this Zephyr vs. Contiki behavior WRT fragmentation and I'll have more to add here soon.

**Comment by nashif** [comment]
**_by Ravi kumar Veeramally:_**

Gustavo, I am working on fixing zephyr fragmentation issue you mentioned above. Issue is simple but fixing in current environment is quite tricky. We do have now BT as bearer. Fragmentation is different with BT. but compression is common. In bt total length of datagram is after compression, but with 15.4 fragmentation datagram size and offset is before compression. I will send you a patch for verification once it is done. Let me know your findings wrt zephyr fragmentation.

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

Zephyr vs. Contiki is indeed broken on the fragmentation: the latter can't reassemble from the former. Contiki vs. itself on the board goes fine. The packet loss on the HW seems to be indeed an (orthogonal, but annoying to the tests in question) issue -- turning off auto-ack on Contiki makes fragmentation a hell to work due to loss. Anyway, our report here is WRT fragmentation and interoperability. We'll have to live with the autoack thing turned on, it seems.

**Comment by nashif** [comment]
**_by Tomasz Bursztyka:_**

Edited the bug fully, so it's title is more relevant etc...

**Comment by nashif** [comment]
**_by Anas Nashif:_**

is this fixed already?

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

I only had an informal patch as feedback for this and was waiting for one entry on gerrit to test for sure. _Ravi kumar Veeramally_ , do we have it there? At least under your name I can't see it.

**Comment by nashif** [comment]
**_by Iván Briano:_**

I'm following Zephyr closely and there has been nothing changing in any of the related code. I don't think we even need to test this to know that it hasn't been fixed yet.

**Comment by nashif** [comment]
**_by Ravi kumar Veeramally:_**

I sent patch to Gustavo for verification almost 3 weeks back, waiting for his feedback on the patch. So that I can send it to gerrit.

**Comment by nashif** [comment]
**_by Iván Briano:_**

Don't do that, or things like this are bound to occur. Post the patch here for everyone to see, test and comment on. Or send it directly to gerrit if it's ready enough, or if you must use email, send it to the mailing list. Any option that keeps everyone in the loop is valid, private emails where the rest of us don't know what's going on are not.

**Comment by nashif** [comment]
**_by Ravi kumar Veeramally:_**

When I discussed with Gustavo, he will verify and gives me some feedback. but he might be occupied with other work, it didn't go further. Anyway I sent to gerrit this morning. Please have a look.

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

I'm indeed swamped with (other) work, sorry. Thus why a WIP-status patch would have to wait for the other TODO. Now we're going to raise the priority on this.

**Comment by nashif** [comment]
**_by Andrei Laperie:_**

We won't do any further work on this problem as part of the stack maintenance.  It will be likely fixed in a new stack.

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

This should have been closed for the current stack as well already. Doing it now.

**Comment by nashif** [comment]
**_by Andrei Laperie:_**

Gustavo, please verify

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

Resolved (FIXED) as it should. Thanks.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

Please reporter verify this.
_Gustavo Lima Chaves_ , as you've verified the issue is fixed, as reporter please change the state to "closed": simply press the "Verify" button then the "Close" button under the title of this issue (on the top of this page). Thanks.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Fixing incorrect priority
